File: LeetCode/Dp.java
import java.util.*;
class Solution{
	void printmatirx(boolean [] dp)
	{
		for (int i=0;i<dp.length ;++i ) {
			if(dp[i])
				System.out.print("1 ");
			else
				System.out.print("0 ");
		}
		System.out.println();
	}
public boolean isMatch(String s, String p) {
    boolean[] match = new boolean[s.length()+1];
    Arrays.fill(match, false);
    match[s.length()] = true;
    for(int i=p.length()-1;i>=0;i--){
        if(p.charAt(i)=='*'){
            for(int j=s.length()-1;j>=0;j--)    match[j] = match[j]||match[j+1]&&(p.charAt(i-1)=='.'||s.charAt(j)==p.charAt(i-1));
            i--;
        }
        else{
            for(int j=0;j<s.length();j++)   match[j] = match[j+1]&&(p.charAt(i)=='.'||p.charAt(i)==s.charAt(j));
            match[s.length()] = false;
        }
        printmatirx(match);
    }
    return match[0];
}
}
public class Dp{
	public static void main(String args[]){
		Solution solution = new Solution();
		String s="a";
		String p="a*a*a*";
		boolean result=solution.isMatch(s,p);
		System.out.println(result); 
	}

}

File: LeetCode/dp.cpp
#include <iostream>
#include <string>
#include <stdlib.h>
using namespace std;
class Solution {
public:
    bool isMatch(string s, string p) {
		int n=s.length();
		int m=p.length();
		bool** dp=new bool*[n+1];
		for(int i=0;i<=n;i++)
		{
			dp[i]=new bool[m+1];
		}
		dp[0][0]=true;
		for(int i=1;i<=n;i++)
		{
			dp[i][0]=false;
		}
		for(int i=1;i<=m;i++)
		{
			dp[0][i]=false;
			if(p[i-1]=='*')
			{
				dp[0][i]=dp[0][i-2];
			}
		}
		
		for(int i=1;i<=n;i++)
		{
			for(int j=1;j<=m;j++)
			{

				dp[i][j]=false;
				if(p[j-1]=='*')
				{
					if(dp[i][j-2]||((s[i-1]==p[j-2]||p[j-2]=='.') && dp[i-1][j]))
						dp[i][j]=true;
				}
				else if(dp[i-1][j-1] && (s[i-1]==p[j-1] || p[j-1]=='.'))
				{
					dp[i][j]=true;
				}

			}
		}
		bool flag=dp[n][m];
		for(int i=0;i<=n;i++){
			for(int j=0;j<=m;j++){
				cout<<dp[i][j]<<" ";
			}
			cout<<endl;
		}

		for(int i=0;i<=n;i++)
		{
			delete [] dp[i];
		} 
		delete[] dp;
		return flag; 
    }
};
int  main(){
	Solution solution;
	string str1="aa";
	string str2="a*q";
	bool result=solution.isMatch(str1,str2);
	if (result) cout<<"True"<<endl;
		else cout<<"False"<<endl;
	system("pause");
	return 0;


}